certbot/certbot/display/util.py

```diff
@@ -46,26 +47,185 @@
+    least one letter each.
+
+    :param str message: question for the user
+    :param str yes_label: Label of the "Yes" parameter
+    :param str no_label: Label of the "No" parameter
+    :param default: default value to return (if one exists)
+    :param str cli_flag: option used to set this value with the CLI
+    :param bool force_interactive: True if it's safe to prompt the user
+        because it won't cause any workflow regressions
+
+    :returns: True for "Yes", False for "No"
+    :rtype: bool
+
+    """
+    return get_display().yesno(message, yes_label=yes_label, no_label=no_label, default=default,
+                               cli_flag=cli_flag, force_interactive=force_interactive)
+
+
+def checklist(message: str, tags: List[str], default: Optional[str] = None,
+              cli_flag: Optional[str] = None,
+              force_interactive: bool = False) -> Tuple[str, List[str]]:
+    """Display a checklist.
+
+    :param str message: Message to display to user
+    :param list tags: `str` tags to select, len(tags) > 0
+    :param default: default value to return (if one exists)
+    :param str cli_flag: option used to set this value with the CLI
+    :param bool force_interactive: True if it's safe to prompt the user
+        because it won't cause any workflow regressions
+
+    :returns: tuple of (`code`, `tags`) where
+        `code` - str display exit code
+        `tags` - list of selected tags
+    :rtype: tuple
+
+    """
+    return get_display().checklist(message, tags, default=default, cli_flag=cli_flag,
+                                   force_interactive=force_interactive)
+
+
+def directory_select(message: str, default: Optional[str] = None, cli_flag: Optional[str] = None,
+                     force_interactive: bool = False) -> Tuple[int, str]:
+    """Display a directory selection screen.
+
+    :param str message: prompt to give the user
+    :param default: default value to return (if one exists)
+    :param str cli_flag: option used to set this value with the CLI
+    :param bool force_interactive: True if it's safe to prompt the user
+        because it won't cause any workflow regressions
+
+    :returns: tuple of the form (`code`, `string`) where
+        `code` - display exit code
+        `string` - input entered by the user
 
     """
-    lines = msg.splitlines()
-    fixed_l = []
+    return get_display().directory_select(message, default=default, cli_flag=cli_flag,
+                                          force_interactive=force_interactive)
+
+
+# The two following functions use "Any" for their parameter/output types. Normally interfaces from
+# certbot.interfaces would be used, but MyPy will not understand their semantic. These interfaces
+# will be removed soon and replaced by ABC classes that will be used also here for type checking.
+# TODO: replace Any by actual ABC classes once available
+
+def get_display() -> Any:
```

Comment:
I wonder if we could and should make this function private? Why should the user of this module ever access the display object instead of using the module level functions? Looking through the code, I found only two use cases:

1. Unit tests where I think we could mock all of `certbot.display.util`. (Since we tell people to run our tests with `pytest`, I'm also interested in doing something like setting up a `context.py` file which creates a simple display object when running tests to remove the need to use mocks here in some cases, however, I think this is out of the scope of this PR.)
2. The code [here](https://github.com/certbot/certbot/blob/c372dd8aeef509c193eb1d4615c39b5860b81293/certbot/certbot/_internal/cli/helpful.py#L68-L74) where `self.notify` is used [here](https://github.com/certbot/certbot/blob/c372dd8aeef509c193eb1d4615c39b5860b81293/certbot/certbot/_internal/cli/helpful.py#L154-L157). I think the only time `self.notify` is used, `zope.component.getUtility` fails because we're parsing arguments before setting up the display and noticing that the user just requested `--help` information which we print and exit. I think we could probably either replace the usage of `self.notify` with `print` or set up a temporary display that can be used until arguments are parsed.

What do you think? Do you think these changes are worth it? If so, I could help make them if you like.

Reply:
Yes I hesitated to make it private. Ideally, we should always access the underlying display object through the module functions. Case 1 is not a big deal, 2 was a little more tricky. My intention was indeed to use a dedicate display object for `self.notify`, but could not find an approach that ensures I did not missed some corner case:
* either we set the display object by `set_display`, but in this case invoking the `HelpArgumentParser` has the side effect to define globally a display without notice
* either we use a local display object but we do not benefit from an existing display that may potentially be set before we create the `HelpArgumentParser` instance.

Now if you say that one or all these situations are not a problem, I can update accordingly the code so that we do not need to use `get_display()`.

Reply:
What do you think about something like:
```
diff --git a/certbot/certbot/_internal/main.py b/certbot/certbot/_internal/main.py
index ed9476c20..c02c2e29f 100644
--- a/certbot/certbot/_internal/main.py
+++ b/certbot/certbot/_internal/main.py
@@ -1511,6 +1511,11 @@ def main(cli_args=None):
     # Some releases of Windows require escape sequences to be enable explicitly
     misc.prepare_virtual_console()
 
+    # HelpfulArgumentParser in prepare_and_parse_args uses display_util so we
+    # create a temporary display that it can use.
+    pre_are_parse_display = display_util.NoninteractiveDisplay(sys.stdout)
+    zope.component.provideUtility(pre_are_parse_display)
+
     # note: arg parser internally handles --help (and exits afterwards)
     args = cli.prepare_and_parse_args(plugins, cli_args)
     config = configuration.NamespaceConfig(args)
```
I'd personally kind of prefer to refactor `HelpfulArgumentParser` so that it doesn't use `display_util`, but I think a change like this is simpler and makes more sense for this PR. This stuff is all internal so we can refactor it later if we really want.

Reply:
For a few lines of code more, what about creating this display object inside `prepare_and_parse_args`, and pass it as a constructor argument to `HelpfulArgumentParser`?

Reply:
That works for me!

Reply:
New update. After looking at the code again, I convinced myself that `self.notify` was always set to a dedicated display object and not a preexisting one. Given that `HelpfulArgumentParser` is internal, I took the simplest approach of removing the branch and always create this dedicated display object during the initialization of an instance of `HelpfulArgumentParser`.

Reply:
That also works for me! At least when `self.notify` is actually used, I agree it's always a dedicated display object.

With that out of the way, I'd personally like to make `get_display` private to remove it from our API. I think in the medium/long term, we should fix up any tests mocking `get_display` and have them mock `certbot.display.util` instead, but I personally think mocking out a private `_get_display` method temporarily may be fine in some cases for the purposes of getting this PR merged.

Reply:
This is done.